wp-includes/class-wp-offline-page-ui.php

```diff
@@ -0,0 +1,327 @@
+			return false;
+		}
+
+		$page_id = wp_insert_post(
+			array(
+				'post_title'   => __( 'Offline Page', 'pwa' ),
+				'post_status'  => 'draft',
+				'post_type'    => 'page',
+				'post_content' => $this->get_default_content(),
+			),
+			true
+		);
+
+		if ( ! is_wp_error( $page_id ) ) {
+			update_option( WP_Offline_Page::OPTION_NAME, $page_id );
+			if ( wp_redirect( admin_url( 'post.php?post=' . $page_id . '&action=edit' ) ) ) {
+				exit;
+			}
+
+			return;
+		}
+
+		add_settings_error(
+			WP_Offline_Page::OPTION_NAME,
+			WP_Offline_Page::OPTION_NAME,
+			__( 'Unable to create the offline page.', 'pwa' ),
+			'error'
+		);
+	}
+
+	/**
+	 * Get the default content for a new offline page.
+	 *
+	 * @todo provide content here for shortcode (classic) or block (Gutenberg)
+	 */
+	protected function get_default_content() {
+		return '';
+	}
+
+	/**
+	 * Add a setting error message when the Offline Page does not exist or is in trash.
+	 *
+	 * @param WP_Post|null|string $offline_page Optional. Instance of the page's `WP_Post` or `null`. Default is an
+	 *                                          empty string. When the default, attempts to look up the offline page.
+	 *
+	 * @return bool Returns true when setting error is added.
+	 */
+	public function add_settings_error( $offline_page = '' ) {
+		if ( '' === $offline_page ) {
+			if ( $this->manager->get_offline_page_id() < 1 ) {
+				return false;
+			}
+
+			$offline_page = get_post( $this->manager->get_offline_page_id() );
+		}
+
+		if ( $this->does_not_exist( $offline_page ) ) {
+			add_settings_error(
+				WP_Offline_Page::OPTION_NAME,
+				WP_Offline_Page::OPTION_NAME,
+				__( 'The current offline page does not exist. Please select or create one.', 'pwa' )
+			);
+
+			return true;
+		}
+
+		if ( $this->is_in_trash_error( $offline_page ) ) {
+			add_settings_error(
+				WP_Offline_Page::OPTION_NAME,
+				WP_Offline_Page::OPTION_NAME,
+				sprintf(
+					/* translators: URL to Pages Trash */
+					__( 'The currently offline page is in the trash. Please select or create one or <a href="%s">restore the current page</a>.', 'pwa' ),
+					'edit.php?post_status=trash&post_type=page'
+				)
+			);
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * When the given post is the offline page, add the state to the given post states.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param array   $post_states An array of post display states.
+	 * @param WP_Post $post        The current post object.
+	 *
+	 * @return array
+	 */
+	public function add_post_state( array $post_states, $post ) {
+		if ( $this->manager->get_offline_page_id() === $post->ID ) {
+			$post_states[] = __( 'Offline Page', 'pwa' );
+		}
+
+		return $post_states;
+	}
+
+	/**
+	 * Gets the configured static pages IDs.
+	 *
+	 * @param bool $join Optional. When true, a comma-delimited list is returned; else, an array is returned.
+	 *
+	 * @return array|string
+	 */
+	protected function get_static_pages_ids( $join = false ) {
```

Comment:
See note below about how `$join` isn't necessary since we can just `implode( ',', ... )` when calling.

Also, I don't think we need `get_static_pages_ids` either. We can just do:

```php
array_filter( $this->manager->get_static_pages() )
```